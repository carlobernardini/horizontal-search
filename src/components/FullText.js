import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { debounce, isFunction } from 'lodash-es';
import { DISPLAY_NAME_FILTER_FULLTEXT } from '../constants';
import Caption from './Caption';
import FilterInfo from './FilterInfo';

class FullText extends Component {
    constructor(props) {
        super(props);
        this.state = {
            derivedValue: props.value,
            value: props.value || ''
        };

        this.dispatchDebounced = debounce(
            this.dispatchToQuery,
            props.dispatchTimeout
        );
        this.input = React.createRef();
        this.dispatchDebouncedWrapper = this.dispatchDebouncedWrapper.bind(this);
        this.dispatchToQuery = this.dispatchToQuery.bind(this);
    }

    componentDidMount() {
        const { autoFocus } = this.props;

        if (!autoFocus || !this.input || !this.input.current) {
            return false;
        }

        this.input.current.focus();
        return true;
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        const { value } = nextProps;
        const { derivedValue } = prevState;

        if (value === null && derivedValue !== value) {
            return {
                derivedValue: value,
                value: ''
            };
        }

        return {
            derivedValue: value
        };
    }

    dispatchDebouncedWrapper() {
        if (!this.dispatchDebounced) {
            return true;
        }
        if (this.dispatchDebounced.cancel) {
            this.dispatchDebounced.cancel();
        }
        this.dispatchDebounced();
        return true;
    }

    dispatchToQuery() {
        const { onUpdate, validateInput } = this.props;
        const { value = '' } = this.state;

        if (isFunction(validateInput) && !validateInput(value)) {
            return false;
        }

        if (!value.trim().length) {
            return onUpdate(null);
        }

        return onUpdate(value);
    }

    renderField() {
        const {
            disabled, maxHeight, maxWidth, multiline, validateInput,
            placeholder, value: derivedValue
        } = this.props;

        const { value = '' } = this.state;

        let fieldProps = {
            className: classNames('orizzonte__filter-fulltext', {
                'orizzonte__filter-fulltext--disabled': disabled,
                'orizzonte__filter-fulltext--invalid': value.length && isFunction(validateInput) && !validateInput(value)
            }),
            disabled,
            onBlur: this.dispatchToQuery,
            onChange: (e) => {
                const { value: val = '' } = e.target;

                if (val === value) {
                    return false;
                }

                this.setState({
                    derivedValue,
                    value: val
                }, this.dispatchDebouncedWrapper);
                return true;
            },
            placeholder,
            ref: this.input,
            value
        };

        if (multiline) {
            fieldProps = {
                ...fieldProps,
                className: classNames(fieldProps.className, 'orizzonte__filter-fulltext--multiline'),
                style: ((mh, mw) => {
                    if (!mh && !mw) {
                        return null;
                    }

                    return {
                        maxHeight: mh ? `${ mh }px` : null,
                        maxWidth: mw ? `${ mw }px` : null
                    };
                })(maxHeight, maxWidth)
            };

            return (
                <textarea
                    { ...fieldProps }
                />
            );
        }

        return (
            <input
                type="text"
                onKeyUp={ (e) => {
                    const key = e.keyCode || e.which;

                    if (key !== 13) {
                        return true;
                    }

                    this.dispatchToQuery();
                    return true;
                }}
                { ...fieldProps }
            />
        );
    }

    render() {
        const { information, label } = this.props;

        return (
            <div
                className="orizzonte__filter"
            >
                <FilterInfo
                    information={ information }
                />
                <Caption>
                    { label }
                </Caption>
                { this.renderField() }
            </div>
        );
    }
}

FullText.displayName = DISPLAY_NAME_FILTER_FULLTEXT;

FullText.propTypes = {
    /** If the textarea should focus when group is expanded
        Only works when group has only one (enabled) fulltext filter */
    autoFocus: PropTypes.bool,
    /** If the textarea should be disabled */
    disabled: PropTypes.bool,
    /** Custom debounce timeout before dispatching the new value to the query object */
    dispatchTimeout: PropTypes.number,
    /** Field name for this filter, to be used in composed query */
    // eslint-disable-next-line
    fieldName: PropTypes.string,
    information: PropTypes.string,
    /** Label for this filter section */
    label: PropTypes.string,
    /** Maximum textarea height (only applicable in multiline mode) */
    maxHeight: PropTypes.number,
    /** Maximum textarea width (only applicable in multiline mode) */
    maxWidth: PropTypes.number,
    /** Whether to render a textarea (true) or input field (false) */
    multiline: PropTypes.bool,
    /** Internal callback for filter update */
    onUpdate: PropTypes.func,
    /** Placeholder text for the input field */
    placeholder: PropTypes.string,
    /** Transforming function or placeholder for group label */
    selectedLabel: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.string
    ]),
    /** Function to validate user input, should return true (valid) or false (invalid) */
    validateInput: PropTypes.func,
    /** Current value for this filter */
    value: PropTypes.string
};

FullText.defaultProps = {
    autoFocus: false,
    disabled: false,
    dispatchTimeout: 300,
    fieldName: null,
    information: null,
    label: null,
    maxHeight: null,
    maxWidth: null,
    multiline: false,
    onUpdate: () => {},
    placeholder: null,
    selectedLabel: '%s',
    validateInput: null,
    value: null
};

export default FullText;
