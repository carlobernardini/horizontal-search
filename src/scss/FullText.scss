@import 'base';

.orizzonte__filter-fulltext {
    width: 100%;
    padding: 5px;
    background-color: #fff;
    border: 1px solid $border-color;
    border-radius: 3px;
    outline: none;
    box-shadow: 0 2px 2px rgba(0, 0, 0, .1);
    transition: border-color .3s;
    box-sizing: border-box;
    height: 32px;
    @extend %fonts;
    font: {
        size: .9em;
    }

    &:focus {
        border-color: $color-a;
        &.orizzonte__filter-fulltext--invalid {
            border-color: $color-e;
        }
    }
    &--disabled {
        cursor: not-allowed;
        background-color: $background-color-muted;
    }
    &--invalid {
        border-color: $color-e;
    }
    &--multiline {
        height: auto;
        min-height: 50px;
        min-width: 190px;
    }
}
