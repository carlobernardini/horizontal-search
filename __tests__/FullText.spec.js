import React from 'react';
import FullText from '../src/components/FullText';

describe('<FullText />', () => {
    jest.useFakeTimers();

    it('should render a fulltext filter', () => {
        const expectedValue = 'some keywords';
        const onUpdate = jest.fn();

        const wrapper = shallow(
            <FullText
                label="Fulltext filter"
                placeholder="This is a textarea placeholder..."
                onUpdate={ onUpdate }
            />
        );

        wrapper.find('.orizzonte__filter-fulltext').simulate('change', {
            target: {
                value: expectedValue
            }
        });

        expect(wrapper.state().value).toBe(expectedValue);
        jest.runAllTimers();
        expect(onUpdate).toHaveBeenCalledWith(expectedValue);
        expect(wrapper).toMatchSnapshot();

        wrapper.find('.orizzonte__filter-fulltext').simulate('change', {
            target: {
                value: ''
            }
        });

        jest.runAllTimers();
        expect(onUpdate).toHaveBeenCalledWith(null);
    });

    it('should render a disabled fulltext filter', () => {
        const wrapper = shallow(
            <FullText
                label="Fulltext filter"
                placeholder="This is a textarea placeholder..."
                disabled
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a multiline fulltext filter with current value and max width / height', () => {
        const wrapper = shallow(
            <FullText
                label="Fulltext filter"
                placeholder="This is a textarea placeholder..."
                value="Current textarea value"
                maxWidth={ 500 }
                maxHeight={ 500 }
                multiline
            />
        );

        expect(wrapper).toMatchSnapshot();

        wrapper.setProps({
            value: 'updated value'
        });

        expect(wrapper.state().derivedValue).toBe('updated value');

        wrapper.setProps({
            value: null
        });

        expect(wrapper.state()).toEqual({
            derivedValue: null,
            value: ''
        });
    });

    it('should render a fulltext filter with input validation', () => {
        const onUpdate = jest.fn();

        const wrapper = shallow(
            <FullText
                label="Fulltext filter"
                placeholder="This is a textarea placeholder..."
                onUpdate={ onUpdate }
                validateInput={ (value) => (
                    !(/[0-9]/g.test(value))
                )}
            />
        );

        wrapper.find('.orizzonte__filter-fulltext').simulate('change', {
            target: {
                value: '123'
            }
        });

        expect(wrapper.state().value).toBe('123');
        jest.runAllTimers();
        expect(wrapper.find('.orizzonte__filter-fulltext').hasClass('orizzonte__filter-fulltext--invalid')).toBe(true);
        expect(onUpdate).not.toHaveBeenCalled();
        expect(wrapper.state().derivedValue).toBeNull();
    });
});
