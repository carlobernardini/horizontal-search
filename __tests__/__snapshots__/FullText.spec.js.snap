// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FullText /> should render a disabled fulltext filter 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Fulltext filter
  </OrizzonteFilterCaption>
  <input
    className="orizzonte__filter-fulltext orizzonte__filter-fulltext--disabled"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="This is a textarea placeholder..."
    type="text"
    value=""
  />
</div>
`;

exports[`<FullText /> should render a fulltext filter 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Fulltext filter
  </OrizzonteFilterCaption>
  <input
    className="orizzonte__filter-fulltext"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="This is a textarea placeholder..."
    type="text"
    value="some keywords"
  />
</div>
`;

exports[`<FullText /> should render a fulltext filter with current value 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <OrizzonteFilterCaption>
    Fulltext filter
  </OrizzonteFilterCaption>
  <input
    className="orizzonte__filter-fulltext"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="This is a textarea placeholder..."
    type="text"
    value="Current textarea value"
  />
</div>
`;
